server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java

```diff
@@ -0,0 +1,233 @@
+
+    private volatile TimeValue monitorInterval;
+    private volatile boolean enabled;
+    // Signals that the health metadata is available, so we can start monitoring.
+    private volatile boolean healthMetadataInitialized;
+    // Ensures that there will no parallel executions of the monitoring process and
+    // simplifies the rescheduling during enabling/disabling or the change of the interval.
+    private final AtomicBoolean inProgress = new AtomicBoolean();
+    // Keeps the latest health state that was successfully reported.
+    private IndividualNodeHealth lastReportedHealth = null;
+
+    public LocalHealthMonitor(Settings settings, ClusterService clusterService, NodeService nodeService, ThreadPool threadPool) {
+        this.threadPool = threadPool;
+        this.monitorInterval = INTERVAL_SETTING.get(settings);
+        this.enabled = HealthNodeTaskExecutor.ENABLED_SETTING.get(settings);
+        this.clusterService = clusterService;
+        this.diskCheck = new DiskCheck(nodeService);
+        clusterService.addListener(this);
+        ClusterSettings clusterSettings = clusterService.getClusterSettings();
+        clusterSettings.addSettingsUpdateConsumer(INTERVAL_SETTING, this::setMonitorInterval);
+        clusterSettings.addSettingsUpdateConsumer(HealthNodeTaskExecutor.ENABLED_SETTING, this::setEnabled);
+    }
+
+    void setMonitorInterval(TimeValue monitorInterval) {
+        this.monitorInterval = monitorInterval;
+        scheduleNowIfEnabled();
+    }
+
+    void setEnabled(boolean enabled) {
+        this.enabled = enabled;
+        scheduleNowIfEnabled();
+    }
+
+    private void scheduleNextRunIfEnabled(TimeValue time) {
+        if (enabled && healthMetadataInitialized) {
+            threadPool.scheduleUnlessShuttingDown(time, ThreadPool.Names.MANAGEMENT, this::monitorHealth);
+        }
+    }
+
+    // Helper method that starts the monitoring without a delay.
+    private void scheduleNowIfEnabled() {
+        scheduleNextRunIfEnabled(new TimeValue(1));
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        // Wait until every node in the cluster is upgraded to 8.4.0 or later
+        if (event.state().nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_4_0)) {
+            // Wait until the health metadata is available in the cluster state
+            if (healthMetadataInitialized == false) {
+                healthMetadataInitialized = HealthMetadata.getFromClusterState(event.state()) != null;
+                scheduleNowIfEnabled();
+            }
+        }
+    }
+
+    // Visible for testing
+    void monitorHealth() {
+        if (inProgress.compareAndSet(false, true)) {
+            ClusterState clusterState = clusterService.state();
+            HealthMetadata healthMetadata = HealthMetadata.getFromClusterState(clusterState);
+            assert healthMetadata != null : "health metadata should have been initialized.";
```

Comment:
Before we schedule any monitoring task we check the flag `healthMetadataInitialized` so theoretically this should hold true. The only case I can think of, is if someone removes the health metadata. But I think we do not give that option programmatically. Would you prefer to have an `if` here?

Reply:
Ah, right. The `HealthMetadataService` is controlled by a different setting - enabled/disabled via `health.node.enabled` so that made me think things can diverge. 
I prefer the explicit checks, but it's a personal preference so feel free to ignore

Reply:
They are all controlled by the same flag `health.node.enabled`. The reason I have done this, is that if the health node is disabled, there is no point to monitor the disk health locally, or to publish health metadata right?

On a second thought, depending on the use case @jbaiera is working on we might want to decouple this and have the health metadata always be published. They might be useful for the coordinator node too.